eth/catalyst/api.go

```diff
@@ -549,83 +561,84 @@ func (api *ConsensusAPI) GetBlobsV1(hashes []common.Hash) ([]*engine.BlobAndProo
 }
 
 // NewPayloadV2 creates an Eth1 block, inserts it in the chain, and returns the status of the chain.
 func (api *ConsensusAPI) NewPayloadV2(params engine.ExecutableData) (engine.PayloadStatusV1, error) {
-	if api.eth.BlockChain().Config().IsCancun(api.eth.BlockChain().Config().LondonBlock, params.Timestamp) {
-		return engine.PayloadStatusV1{Status: engine.INVALID}, engine.InvalidParams.With(errors.New("can't use newPayloadV2 post-cancun"))
-	}
-	if api.eth.BlockChain().Config().LatestFork(params.Timestamp) == forks.Shanghai {
-		if params.Withdrawals == nil {
-			return engine.PayloadStatusV1{Status: engine.INVALID}, engine.InvalidParams.With(errors.New("nil withdrawals post-shanghai"))
-		}
-	} else {
-		if params.Withdrawals != nil {
-			return engine.PayloadStatusV1{Status: engine.INVALID}, engine.InvalidParams.With(errors.New("non-nil withdrawals pre-shanghai"))
-		}
-	}
-	if params.ExcessBlobGas != nil {
-		return engine.PayloadStatusV1{Status: engine.INVALID}, engine.InvalidParams.With(errors.New("non-nil excessBlobGas pre-cancun"))
+	var (
+		isCancun     = api.eth.BlockChain().Config().IsCancun(api.eth.BlockChain().Config().LondonBlock, params.Timestamp)
+		isInShanghai = api.isLatestForkAt(params.Timestamp, forks.Shanghai)
+	)
+
+	checks := []param{
+		{isCancun, "can't use newPayloadV2 post-cancun"},
+		{isInShanghai && params.Withdrawals == nil, "nil withdrawals post-shanghai"},
+		{!isInShanghai && params.Withdrawals != nil, "non-nil withdrawals pre-shanghai"},
+		{params.ExcessBlobGas != nil, "non-nil excessBlobGas pre-cancun"},
+		{params.BlobGasUsed != nil, "non-nil blobGasUsed pre-cancun"},
 	}
-	if params.BlobGasUsed != nil {
-		return engine.PayloadStatusV1{Status: engine.INVALID}, engine.InvalidParams.With(errors.New("non-nil blobGasUsed pre-cancun"))
+	if err := verifyConditions(checks); err != nil {
+		return engine.PayloadStatusV1{Status: engine.INVALID}, err
 	}
+
 	return api.newPayload(params, nil, nil, nil, false)
 }
 
-// NewPayloadV3 creates an Eth1 block, inserts it in the chain, and returns the status of the chain.
-func (api *ConsensusAPI) NewPayloadV3(params engine.ExecutableData, versionedHashes []common.Hash, beaconRoot *common.Hash) (engine.PayloadStatusV1, error) {
-	if params.Withdrawals == nil {
-		return engine.PayloadStatusV1{Status: engine.INVALID}, engine.InvalidParams.With(errors.New("nil withdrawals post-shanghai"))
-	}
-	if params.ExcessBlobGas == nil {
-		return engine.PayloadStatusV1{Status: engine.INVALID}, engine.InvalidParams.With(errors.New("nil excessBlobGas post-cancun"))
-	}
-	if params.BlobGasUsed == nil {
-		return engine.PayloadStatusV1{Status: engine.INVALID}, engine.InvalidParams.With(errors.New("nil blobGasUsed post-cancun"))
+type param struct {
+	condition bool
+	message   string
+}
+
+func verifyConditions(checks []param) error {
+	for _, check := range checks {
+		if check.condition {
+			return engine.InvalidParams.With(errors.New(check.message))
+		}
 	}
+	return nil
+}
```

Comment:
This structure feels weird to me. I think I would prefer to retain the current in line if blocks, but simplify the return value with something like `return newParamsError(msg string)`